File: problems/2197_replace_non_coprime_numbers_in_array/2197_replace_non_coprime_numbers_in_array.py
class Solution:
    def replaceNonCoprimes(self, nums: List[int]) -> List[int]:
        stack = []
        for num in nums:
            stack.append(num)
            while len(stack) > 1 and gcd(stack[-1], stack[-2]) > 1:
                y = stack.pop()
                z = stack.pop()
                stack.append(lcm(y, z))
        return stack

File: problems/2197_replace_non_coprime_numbers_in_array/2197_replace_non_coprime_numbers_in_array.java
class Solution {
    public List<Integer> replaceNonCoprimes(int[] nums) {
        List<Integer> stack = new ArrayList<>();
        for (int num : nums) {
            int cur = num;
            while (!stack.isEmpty()) {
                int last = stack.get(stack.size() - 1);
                int g = gcd(last, cur);
                if (g == 1)
                    break;
                stack.remove(stack.size() - 1);
                cur = lcm(last, cur);
            }
            stack.add(cur);
        }
        return stack;
    }

    private int gcd(int a, int b) {
        return b == 0 ? a : gcd(b, a % b);
    }

    private int lcm(int a, int b) {
        return a / gcd(a, b) * b;
    }
}